### PR TITLE
Implement recovery endpoint

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Implement recovery endpoint. Handles launching a recovered execution and appropriately saving its execution lineage in the data store.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1151

## Follow-up issue
_NA_

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -16,10 +16,10 @@ require (
 	github.com/coreos/go-oidc v2.2.1+incompatible
 	github.com/dgrijalva/jwt-go v3.2.0+incompatible
 	github.com/evanphx/json-patch v4.9.0+incompatible
-	github.com/flyteorg/flyteidl v0.19.5
-	github.com/flyteorg/flyteplugins v0.5.56
-	github.com/flyteorg/flytepropeller v0.12.9
-	github.com/flyteorg/flytestdlib v0.3.22
+	github.com/flyteorg/flyteidl v0.19.14
+	github.com/flyteorg/flyteplugins v0.5.59
+	github.com/flyteorg/flytepropeller v0.12.14-0.20210715214519-2260e0661cab
```
Comment (katrogan) [inline go.mod:21]: will update before merging

File: pkg/manager/impl/execution_manager.go [hunk 2]
```diff
@@ -901,6 +905,57 @@ func (m *ExecutionManager) RelaunchExecution(
...
+	}, requestedAt)
+	if err != nil {
+		return nil, err
+	}
+	executionModel.SourceExecutionID = existingExecutionModel.ID
+	workflowExecutionIdentifier, err := m.createExecutionModel(ctx, executionModel)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debugf(ctx, "Successfully relaunched [%+v] as [%+v]", request.Id, workflowExecutionIdentifier)
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:953]:
```suggestion
	logger.Debugf(ctx, "Successfully relaunched [%+v] as [%+v] in recover mode", request.Id, workflowExecutionIdentifier)
```

File: pkg/rpc/adminservice/execution.go [hunk 3]
```diff
@@ -66,6 +66,31 @@ func (m *AdminService) RelaunchExecution(
...
+func (m *AdminService) RecoverExecution(
+	ctx context.Context, request *admin.ExecutionRecoverRequest) (*admin.ExecutionCreateResponse, error) {
+	defer m.interceptPanic(ctx, request)
+	requestedAt := time.Now()
+	if request == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Incorrect request, nil requests not allowed")
+	}
+	var response *admin.ExecutionCreateResponse
+	var err error
+	m.Metrics.executionEndpointMetrics.relaunch.Time(func() {
```
Comment (EngHabu) [inline pkg/rpc/adminservice/execution.go:78]: relaunch? should we have an additional metric for this?

Comment (katrogan) [reply pkg/rpc/adminservice/execution.go:78]: done

File: pkg/manager/impl/execution_manager.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: done

**Comment by katrogan** [comment]
PTAL @kumare3

File: pkg/rpc/adminservice/execution.go (hunk 3 above)
Comment (kumare3) [reply pkg/rpc/adminservice/execution.go:78]: its still relaunch?

File: pkg/manager/interfaces/execution.go [hunk 4]
```diff
@@ -13,6 +13,8 @@ type ExecutionInterface interface {
 		*admin.ExecutionCreateResponse, error)
 	RelaunchExecution(ctx context.Context, request admin.ExecutionRelaunchRequest, requestedAt time.Time) (
 		*admin.ExecutionCreateResponse, error)
+	RecoverExecution(ctx context.Context, request admin.ExecutionRecoverRequest, requestedAt time.Time) (
```
Comment (kumare3) [inline pkg/manager/interfaces/execution.go:16]: I think we should add extensive documentation at the interface level to explain what each method does. Lets start with recover

File: pkg/rpc/adminservice/execution.go (hunk 3 above)
Comment (katrogan) [reply pkg/rpc/adminservice/execution.go]: done for real, thanks

File: pkg/manager/interfaces/execution.go (hunk 4 above)
Comment (katrogan) [reply pkg/manager/interfaces/execution.go:19]: done

**Review by pmahindrakar-oss — APPROVED** [review]

**Pull request merged**